relayer/crates/starknet-relayer/src/contexts/builder.rs

```diff
@@ -120,6 +122,56 @@ impl ChainBuilder<StarknetBuilder, Index<1>> for StarknetBuildComponents {
     }
 }
 
+#[cgp_provider(RelayBuilderComponent)]
+impl RelayBuilder<StarknetBuilder, Index<0>, Index<1>> for StarknetBuildComponents {
+    async fn build_relay(
+        build: &StarknetBuilder,
+        _index: PhantomData<(Index<0>, Index<1>)>,
+        src_chain_id: &ChainId,
+        dst_chain_id: &ChainId,
+        src_client_id: &ClientId,
+        dst_client_id: &ClientId,
+    ) -> Result<StarknetToCosmosRelay, HermesError> {
+        let src_chain = build.build_chain(src_chain_id).await?;
+
+        let dst_chain = build.cosmos_builder.build_chain(dst_chain_id).await?;
+
+        Ok(
+            build.build_starknet_to_cosmos_relay(
+                src_chain,
+                dst_chain,
+                src_client_id,
+                dst_client_id,
+            ),
+        )
+    }
+}
+
+#[cgp_provider(RelayBuilderComponent)]
+impl RelayBuilder<StarknetBuilder, Index<1>, Index<0>> for StarknetBuildComponents {
+    async fn build_relay(
+        build: &StarknetBuilder,
+        _index: PhantomData<(Index<1>, Index<0>)>,
+        src_chain_id: &ChainId,
+        dst_chain_id: &ChainId,
+        src_client_id: &ClientId,
+        dst_client_id: &ClientId,
+    ) -> Result<CosmosToStarknetRelay, HermesError> {
+        let src_chain = build.cosmos_builder.build_chain(dst_chain_id).await?;
+
+        let dst_chain = build.build_chain(src_chain_id).await?;
+
+        Ok(
+            build.build_cosmos_to_starknet_relay(
+                src_chain,
+                dst_chain,
+                src_client_id,
+                dst_client_id,
+            ),
+        )
+    }
+}
```

Comment:
does this fixes the ordered-pair type issue in `start starknet-with-cosmos` and `start cosmos-with-starknet`?

Also, wanted to confirm if these subcommands are essentially equivalent with each other.

Reply:
> Also, wanted to confirm if these subcommands are essentially equivalent with each other.

The relay[1, 0] builder is mainly used in the `create cosmos-client` CLI, which creates a Starknet client on Cosmos. For the create connection and channel commands, we currently simply hardcode to require the first chain to always be Starknet to simplify the implementation and interface.

Reply:
> So we have to rely on the caller to explicitly disambiguate this.

So the both order works. but Hermes-SDK doesn't have the support for chain type validation.

so users may mix up types and Hermes-SDK will try to relay without proper messages and fail.

did I understand correctly?

Reply:
If you mix up the order, you would just get an error. Because the relayer would for example try to use a Starknet chain ID to construct a Cosmos chain, which it would then fail to find the chain from the config.